win32/devdoc/file_win32_requirements.md

```diff
@@ -0,0 +1,166 @@
+
+**SRS_FILE_WIN32_43_041: [** `file_create` shall allocate a `FILE_HANDLE`. **]**
+
+**SRS_FILE_WIN32_43_002: [** `file_create` shall call `SetFileCompletionNotificationModes` to disable calling the completion port when an async operations finishes synchrounously. **]**
+
+**SRS_FILE_WIN32_43_003: [** `file_create` shall initialize a threadpool environment. **]**
+
+**SRS_FILE_WIN32_43_004: [** `file_create` shall obtain a `PTP_POOL` struct by calling `execution_engine_win32_get_threadpool` on `execution_engine`. **]**
+
+**SRS_FILE_WIN32_43_005: [** `file_create` shall register the threadpool environment with `ptpPool` **]**
+
+**SRS_FILE_WIN32_43_006: [** `file_create` shall create a cleanup group. **]**
+
+**SRS_FILE_WIN32_43_007: [** `file_create` shall register the cleanup group with the threadpool environment. **]**
+
+**SRS_FILE_WIN32_43_033: [** `file_create` shall create a threadpool io and use `on_file_io_complete_win32` as a callback. **]**
+
+**SRS_FILE_WIN32_43_008: [** If there are any failures, `file_create` shall return `NULL`. **]**
+
+**SRS_FILE_WIN32_43_009: [** `file_create` shall succeed and return the file handle of the opened file. **]**
+
+## file_destroy
+
+```c
+MOCKABLE_FUNCTION(, void, file_destroy, FILE_HANDLE, handle);
+```
+
+**SRS_FILE_WIN32_43_042: [** If `handle` is `NULL`, `file_destroy` shall return. **]**
+
+**SRS_FILE_WIN32_43_011: [** `file_destroy` shall wait for all I/O to complete. **]**
+
+**SRS_FILE_WIN32_43_012: [** `file_destroy` shall close the cleanup group. **]**
+
+**SRS_FILE_WIN32_43_013: [** `file_destroy` shall destroy the environment. **]**
+
+**SRS_FILE_WIN32_43_015: [** `file_destroy` shall close the threadpool IO. **]**
+
+**SRS_FILE_WIN32_43_016: [** `file_destroy` shall call `CloseHandle` on `handle` **]**
+
+**SRS_FILE_WIN32_43_042: [** `file_destroy` shall free the `FILE_HANDLE`. **]**
+
+
+## file_write_async
+
+```c
+MOCKABLE_FUNCTION_WITH_RETURNS(, FILE_WRITE_ASYNC_RESULT, file_write_async, FILE_HANDLE, handle, CONSTBUFFER_HANDLE, source, uint64_t, position, FILE_WRITE_CB, user_callback, void*, user_context)(0, MU_FAILURE);
+```
+
+**SRS_FILE_WIN32_43_043: [** If `handle` is `NULL` then `file_write_async` shall fail and return `FILE_WRITE_ASYNC_INVALID_ARGS`. **]**
+
+**SRS_FILE_WIN32_43_044: [** If `source` is `NULL` then `file_write_async` shall fail and return `FILE_WRITE_ASYNC_INVALID_ARGS`. **]**
+
+**SRS_FILE_WIN32_43_045: [** If `user_callback` is `NULL` then `file_write_async` shall fail and return `FILE_WRITE_ASYNC_INVALID_ARGS`. **]**
+
+**SRS_FILE_WIN32_43_017: [** `file_write_async` shall call `StartThreadpoolIo`. **]**
+
+**SRS_FILE_WIN32_43_018: [** `file_write_async` shall allocate a context to store `handle`, write as the type of asyncronous operation, `source`, `user_callback` and `user_context. **]**
+
+**SRS_FILE_WIN32_43_020: [** `file_write_async` shall populate the `ov` field of the `FILE_WIN32_IO` using `position`. **]**

```

Comment:
>FILE_WIN32_IO [](start = 86, length = 13)

whose field?

Reply:
`FILE_WIN32_IO` is defined in file_win32.c

---
In reply to: [450491475](https://github.com/Azure/azure-c-pal/pull/7#discussion_r450491475) [](ancestors = 450491475)

Reply:
Specs should have all the information without needing to look at code, so if you a referencing some internal struct in specs (because it's easier/more clear), then you should show that definition somewhere in the spec (maybe another code section after the Exposed API)

---
In reply to: [451684793](https://github.com/Azure/azure-c-pal/pull/7#discussion_r451684793) [](ancestors = 451684793,450491475)

Reply:
Changed the spec to remove implementation details.

---
In reply to: [451691467](https://github.com/Azure/azure-c-pal/pull/7#discussion_r451691467) [](ancestors = 451691467,451684793,450491475)